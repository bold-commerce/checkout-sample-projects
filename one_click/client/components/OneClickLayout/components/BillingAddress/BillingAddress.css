.FieldSet--BillingAddress .RadioButton .stx-field {
  position: unset;
}
.FieldSet--BillingAddress .RadioButton .stx-radio {
  position: unset;
}

.FieldSet--BillingAddress.FieldSet {
  display: flex;
  flex-direction: column;
  margin: 0;
}

.FieldSet--BillingAddress .RadioButton {
  margin-bottom: 0;
  font-style: normal;
  font-weight: normal;
  font-size: 14px;
  line-height: 20px;
  display: flex;
  align-items: center;
  padding-left: 0;
  cursor: pointer;
}

.FieldSet--BillingAddress .FieldSet__Content {
  display: grid;
  grid-auto-flow: row;
  grid-gap: 10px;
}

.FieldSet--BillingAddress .RadioButton .stx-checkbox-field__checkbox {
  position: relative;
  left: 0;
  top: 0;
  margin-right: 10px;
}

.FieldSet--BillingAddress .RadioButton .stx-radio-field__label {
  position: absolute;
  left: 0;
  height: 100%;
  top: 0;
  padding-left: 45px;
  display: flex;
  flex-direction: column;
  justify-content: center;
  cursor: pointer;
  width: 100%;
}
