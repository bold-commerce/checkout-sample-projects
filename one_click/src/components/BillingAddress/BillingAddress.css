.FieldSet--BillingAddress .RadioButton .stx-field {
  position: unset;
}
.FieldSet--BillingAddress .RadioButton .stx-radio {
  position: unset;
}

.FieldSet--BillingAddress.FieldSet {
  display: flex;
  flex-direction: column;
  margin: 0;
}

.FieldSet--BillingAddress .RadioButton {
  margin-bottom: 0;
  font-style: normal;
  font-weight: normal;
  font-size: 14px;
  line-height: 20px;
  display: flex;
  align-items: center;
  padding-left: 0;
  cursor: pointer;
}

.FieldSet--BillingAddress .FieldSet__Content .stx-checkbox-field__div {
  display: flex;
}

.FieldSet--BillingAddress .RadioButton .stx-checkbox-field__checkbox {
  position: relative;
  margin-right: 10px;
}

.FieldSet--BillingAddress .RadioButton .stx-checkbox-field__label {
  height: 100%;
  cursor: pointer;
  width: 100%;
  color: #000;
}

.Billing__Title {
  font-size: 14px;
  font-weight: bold;
  line-height: 20px;
  color: rgba(0, 0, 0, .9);
  margin-bottom: 4px;
}