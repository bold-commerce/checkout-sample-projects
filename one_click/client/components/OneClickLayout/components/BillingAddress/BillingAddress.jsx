/* eslint-disable react/forbid-prop-types */
import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { CheckboxField } from '@boldcommerce/stacks-ui';
import { Address } from '../Address';
import { useBillingAddress, useBillingSameAsShipping, useCountryInfo } from '@boldcommerce/checkout-react-components';
import './BillingAddress.css';

export const BillingAddress = ({
  billingAddress, billingSameAsShipping, submitBillingAddress, setBillingSameAsShipping, onChange, requiredAddressFields,
}) => {
  const [address, setAddress] = useState(billingAddress);
  const { data } = useCountryInfo(address);
  const{
    countries,
    provinces,
    showProvince,
    showPostalCode,
    provinceLabel,
  } = data;
  const [loading, setLoading] = useState(false);
  const [errors, setErrors] = useState(null);

  let provincePlaceholder = provinceLabel;

  // TODO: replace with languages config file
  if (provinceLabel === 'state_territory') {
    provincePlaceholder = 'state/territory';
  }

  const handleSubmit = async(data) => {
    setLoading(true);
    try {
      await submitBillingAddress(data);
      setErrors(null);
    } catch(e) {
      setErrors(e.body.errors);
    }
    setLoading(false);
  };

  const handleSetBillingSameAsShipping = async(data) => {
    setLoading(true);
    try {
      await setBillingSameAsShipping(data)
      setErrors(null);
    } catch(e) {
      setErrors(e.body.errors);
    }
    setLoading(false);
  };

  return (
    <section className="FieldSet FieldSet--BillingAddress">
      <div className="FieldSet__Header">
        <h3>Billing address</h3>
      </div>
      <div className="FieldSet__Content">
        <CheckboxField
          label="Same as shipping address"
          name="billing-address"
          className="RadioButton"
          value="SAME_AS_SHIPPING_ADDRESS"
          checked={billingSameAsShipping}
          disabled={loading}
          onChange={() => handleSetBillingSameAsShipping(!billingSameAsShipping)}
        />
        { billingSameAsShipping ? null
          : (
            <Address
              address={address}
              onChange={(data) => setAddress((prevAddress) => ({
                ...prevAddress,
                ...data,
              }))}
              errors={errors}
              countries={countries}
              provinces={provinces}
              showPostalCode={showPostalCode}
              showProvince={showProvince}
              provinceLabel={provincePlaceholder}
              submit={() => handleSubmit(address)}
              requiredAddressFields={requiredAddressFields}
            />
          )}
      </div>
    </section>
  );
};

BillingAddress.propTypes = {
  billingAddress: PropTypes.any,
  countryInfo: PropTypes.array,
  billingAddressErrors: PropTypes.object,
  billingSameAsShipping: PropTypes.bool,
  submitBillingAddress: PropTypes.func,
  setBillingSameAsShipping: PropTypes.func,
  onChange: PropTypes.func,
  requiredAddressFields: PropTypes.array,
};

const MemoizedBillingAddress = React.memo(BillingAddress);

const BillingAddressContainer = ({ requiredAddressFields }) => {
  const { data: billingAddress, submitBillingAddress } = useBillingAddress();
  const { data: billingSameAsShipping, setBillingSameAsShipping } = useBillingSameAsShipping(); 

  return (
    <MemoizedBillingAddress
      billingAddress={billingAddress}
      billingSameAsShipping={billingSameAsShipping}
      submitBillingAddress={submitBillingAddress}
      setBillingSameAsShipping={setBillingSameAsShipping}
      requiredAddressFields={requiredAddressFields}
    />
  );
};

BillingAddressContainer.propTypes = {
  requiredAddressFields: PropTypes.array,
};

export default BillingAddressContainer;
