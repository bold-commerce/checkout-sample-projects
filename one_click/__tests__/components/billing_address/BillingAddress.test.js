import React from "react";
import { render } from '@testing-library/react';
import { BillingAddress } from '../../../src/components/BillingAddress/BillingAddress';
import {
    exampleAddress as MOCKexampleAddress,
    countries as MOCKcountries,
    caProvinces as MOCKprovinces
} from '../../utils/addressHelpers';

jest.mock('@boldcommerce/checkout-react-components', () => ({
    ...jest.requireActual('@boldcommerce/checkout-react-components'),
    useCountryInfo: () => ({
        data: {
            countries: MOCKcountries,
            provinceLabel: "province",
            provinces: MOCKprovinces,
            showPostalCode: true,
            showProvince: true
        }
    })
}))

describe('BillingAddress', () => {
    test('render BillingAddress component with the same billing address', () => {
        const { asFragment } = render(
            <BillingAddress
                billingAddress={MOCKexampleAddress}
                billingSameAsShipping={false}
            />
        );
        expect(asFragment()).toMatchSnapshot();
    })

    test('render BillingAddress component with differnt billing address', () => {
        const { asFragment } = render(
            <BillingAddress
                billingAddress={MOCKexampleAddress}
                billingSameAsShipping
            />
        );
        expect(asFragment()).toMatchSnapshot();
    })
})