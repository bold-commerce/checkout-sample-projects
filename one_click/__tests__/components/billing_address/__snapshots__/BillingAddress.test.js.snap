// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BillingAddress render BillingAddress component with differnt billing address 1`] = `
<DocumentFragment>
  <section
    class="FieldSet FieldSet--BillingAddress"
  >
    <div
      class="FieldSet__Header"
    >
      <h3>
        Billing address
      </h3>
    </div>
    <div
      class="FieldSet__Content"
    >
      <div
        class="stx-field stx-checkbox-field__field RadioButton"
      >
        <div
          class="stx-checkbox-field__div"
        >
          <span
            class="stx-checkbox stx-checkbox--is-checked stx-checkbox-field__checkbox"
          >
            <input
              checked=""
              class="stx-checkbox__input"
              id="stxField12"
              name="billing-address"
              type="checkbox"
              value="SAME_AS_SHIPPING_ADDRESS"
            />
            <span
              class="stx-checkbox__mount"
            />
          </span>
          <label
            class="stx-checkbox-field__label"
            for="stxField12"
          >
            Same as shipping address
          </label>
        </div>
      </div>
    </div>
  </section>
</DocumentFragment>
`;

exports[`BillingAddress render BillingAddress component with the same billing address 1`] = `
<DocumentFragment>
  <section
    class="FieldSet FieldSet--BillingAddress"
  >
    <div
      class="FieldSet__Header"
    >
      <h3>
        Billing address
      </h3>
    </div>
    <div
      class="FieldSet__Content"
    >
      <div
        class="stx-field stx-checkbox-field__field RadioButton"
      >
        <div
          class="stx-checkbox-field__div"
        >
          <span
            class="stx-checkbox stx-checkbox-field__checkbox"
          >
            <input
              class="stx-checkbox__input"
              id="stxField1"
              name="billing-address"
              type="checkbox"
              value="SAME_AS_SHIPPING_ADDRESS"
            />
            <span
              class="stx-checkbox__mount"
            />
          </span>
          <label
            class="stx-checkbox-field__label"
            for="stxField1"
          >
            Same as shipping address
          </label>
        </div>
      </div>
      <div
        class="FieldSet--Address"
      >
        <div
          class="FieldGroup"
        >
          <div
            class="stx-field stx-field--with-input Field Field--FirstName"
          >
            <div
              class="stx-input"
            >
              <input
                class="stx-input__input-element"
                id="stxField2"
                name="first_name"
                placeholder="First name (optional)"
                type="text"
                value="John"
              />
            </div>
          </div>
          <div
            class="stx-field stx-field--with-input Field Field--LastName"
          >
            <div
              class="stx-input"
            >
              <input
                class="stx-input__input-element"
                id="stxField3"
                name="last_name"
                placeholder="Last name (optional)"
                type="text"
                value="Doe"
              />
            </div>
          </div>
        </div>
        <div
          class="FieldGroup"
        >
          <div
            class="stx-field stx-field--with-input Field Field--Company"
          >
            <div
              class="stx-input"
            >
              <input
                class="stx-input__input-element"
                id="stxField4"
                name="business_name"
                placeholder="Company (optional)"
                type="text"
                value="Bold Commerce"
              />
            </div>
          </div>
        </div>
        <div
          class="FieldGroup"
        >
          <div
            class="stx-field stx-field--with-input Field Field--Address"
          >
            <div
              class="stx-input"
            >
              <input
                class="stx-input__input-element"
                id="stxField5"
                name="address_line_1"
                placeholder="Address (optional)"
                type="text"
                value="123 Fake Street"
              />
            </div>
          </div>
          <div
            class="stx-field stx-field--with-input Field Field--Address2"
          >
            <div
              class="stx-input"
            >
              <input
                class="stx-input__input-element"
                id="stxField6"
                name="address_line_2"
                placeholder="Apt, suite, etc."
                type="text"
                value="Unit 12"
              />
            </div>
          </div>
        </div>
        <div
          class="FieldGroup"
        >
          <div
            class="stx-field stx-field--with-input Field Field--City"
          >
            <div
              class="stx-input"
            >
              <input
                class="stx-input__input-element"
                id="stxField7"
                name="city"
                placeholder="City (optional)"
                type="text"
                value="Winnipeg"
              />
            </div>
          </div>
        </div>
        <div
          class="FieldGroup FieldGroup__Country"
        >
          <div
            class="stx-field stx-field--with-select SelectField Field--Country"
          >
            <div
              class="stx-select stx-select--with-placeholder"
            >
              <select
                class="stx-select__select-element"
                id="stxField8"
              >
                <option
                  hidden=""
                  value="___PLACEHOLDER___"
                >
                  Select a country
                </option>
                <option
                  value="CA"
                >
                  Canada
                </option>
                <option
                  value="US"
                >
                  United States
                </option>
              </select>
              <span
                class="stx-select__icon"
              >
                <svg
                  class="stx-select-icon__icon"
                  viewBox="0 0 24 24"
                >
                  <path
                    class="stx-select-icon__path"
                    d="M3.75261265,6.34153452 C3.38894463,5.92588561 2.75718344,5.88374677 2.34153452,6.24741478 C1.92588561,6.6110828 1.88374677,7.24284399 2.24741478,7.65849291 L11.2524148,18.2364929 C11.6508167,18.6918404 12.3591606,18.6918563 12.757583,18.2365268 L21.763583,7.6585268 C22.1272697,7.24289426 22.0851593,6.61113117 21.6695268,6.24744443 C21.2538943,5.8837577 20.6221312,5.9258681 20.2584444,6.34150064 L12.0050479,16.0594019 L3.75261265,6.34153452 Z"
                    fill="#6B6B6B"
                  />
                </svg>
              </span>
            </div>
          </div>
          <div
            class="stx-field stx-field--with-select SelectField Field--Province"
          >
            <div
              class="stx-select stx-select--with-placeholder"
            >
              <select
                class="stx-select__select-element"
                id="stxField9"
              >
                <option
                  hidden=""
                  value="___PLACEHOLDER___"
                >
                  Select a province
                </option>
                <option
                  value="AB"
                >
                  Alberta
                </option>
                <option
                  value="BC"
                >
                  British Columbia
                </option>
                <option
                  value="MB"
                >
                  Manitoba
                </option>
                <option
                  value="NB"
                >
                  New Brunswick
                </option>
                <option
                  value="NL"
                >
                  Newfoundland and Labrador
                </option>
                <option
                  value="NT"
                >
                  Northwest Territories
                </option>
                <option
                  value="NS"
                >
                  Nova Scotia
                </option>
                <option
                  value="NU"
                >
                  Nunavut
                </option>
                <option
                  value="ON"
                >
                  Ontario
                </option>
                <option
                  value="PE"
                >
                  Prince Edward Island
                </option>
                <option
                  value="QC"
                >
                  Quebec
                </option>
                <option
                  value="SK"
                >
                  Saskatchewan
                </option>
                <option
                  value="YT"
                >
                  Yukon Territory
                </option>
              </select>
              <span
                class="stx-select__icon"
              >
                <svg
                  class="stx-select-icon__icon"
                  viewBox="0 0 24 24"
                >
                  <path
                    class="stx-select-icon__path"
                    d="M3.75261265,6.34153452 C3.38894463,5.92588561 2.75718344,5.88374677 2.34153452,6.24741478 C1.92588561,6.6110828 1.88374677,7.24284399 2.24741478,7.65849291 L11.2524148,18.2364929 C11.6508167,18.6918404 12.3591606,18.6918563 12.757583,18.2365268 L21.763583,7.6585268 C22.1272697,7.24289426 22.0851593,6.61113117 21.6695268,6.24744443 C21.2538943,5.8837577 20.6221312,5.9258681 20.2584444,6.34150064 L12.0050479,16.0594019 L3.75261265,6.34153452 Z"
                    fill="#6B6B6B"
                  />
                </svg>
              </span>
            </div>
          </div>
          <div
            class="stx-field stx-field--with-input Field Field--Postal_Code"
          >
            <div
              class="stx-input"
            >
              <input
                class="stx-input__input-element"
                id="stxField10"
                name="postal"
                placeholder="Postal code / ZIP"
                type="text"
                value="R3A1B3"
              />
            </div>
          </div>
        </div>
        <div
          class="FieldGroup"
        >
          <div
            class="stx-field stx-field--with-input Field Field--Phone"
          >
            <div
              class="stx-input"
            >
              <input
                class="stx-input__input-element"
                id="stxField11"
                name="phone_number"
                placeholder="Phone (optional)"
                type="tel"
                value="204-123-4567"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
